src/ConstantOptimization.jl

```diff
@@ -93,38 +108,60 @@ function _optimize_constants(
         member.birth = get_birth_order(; deterministic=options.deterministic)
         num_evals += eval_fraction
     else
+        # Reset to original state
         set_scalar_constants!(member.tree, x0, refs)
     end
 
     return member, num_evals
 end
 
-struct Evaluator{N<:AbstractExpression,R,D<:Dataset,O<:AbstractOptions} <: Function
-    tree::N
-    refs::R
+struct EvaluatorContext{D<:Dataset,O<:AbstractOptions} <: Function
     dataset::D
     options::O
 end
+function (c::EvaluatorContext)(tree; regularization=false)
+    return eval_loss(tree, c.dataset, c.options; regularization)
+end
+
+struct Evaluator{N<:AbstractExpression,R,C<:EvaluatorContext} <: Function
+    tree::N
+    refs::R
+    ctx::C
+end
 function (e::Evaluator)(x::AbstractVector; regularization=false)
     set_scalar_constants!(e.tree, x, e.refs)
-    return eval_loss(e.tree, e.dataset, e.options; regularization)
+    return e.ctx(e.tree; regularization)
 end
 
-struct GradEvaluator{F<:Evaluator,AD<:Union{Nothing,AbstractADType},EX} <: Function
-    f::F
+struct GradEvaluator{E<:Evaluator,AD<:Union{Nothing,AbstractADType},PR,EX} <: Function
+    e::E
+    prep::PR
     backend::AD
     extra::EX
 end
-GradEvaluator(f::F, backend::AD) where {F,AD} = GradEvaluator(f, backend, nothing)
+@unstable function GradEvaluator(e::Evaluator, backend)
+    prep = isnothing(backend) ? nothing : _cached_prep(e.ctx, backend, e.tree)
+    return GradEvaluator(e, prep, backend, nothing)
+end
+
+const CachedPrep = PerTaskCache{Dict{UInt,Any}}()
+
+@unstable function _cached_prep(ctx, backend, example_tree)
+    # We avoid hashing on the tree _value_ because it should not
+    # affect the prep. We want to cache as much as possible!
+    key = hash((ctx, backend, typeof(example_tree)))
+    stable_get!(CachedPrep[], key) do
+        prepare_gradient(ctx, backend, example_tree)
+    end
+end
```

Comment:
@yebai @gdalle is there a robust way to cache the output of `prepare_gradient`? I was trying to figure out what sorts of things would invalidate said cache.

Basically I don't want to re-compute the prep for a new expression instance, since it's slow. So I cache the prep in `Base.task_local_storage()`, at a key based on (1) the dataset, (2) the backend, and (3) the expression's _type_.

But wasn't sure if this is going to break things or not.

Perhaps I can also cache based on the dataset's type, rather than the dataset's value?

Reply:
Assuming this is ok